country-a-service/authentication/src/main/java/dk/sundhedsdatastyrelsen/ncpeh/authentication/CachedAuthenticationService.java

```diff
@@ -0,0 +1,135 @@
+                .maximumSize(10000)
+                .build();
+    }
+
+    public CachedAuthenticationService(AuthenticationServiceInterface service, String issuer) {
+        this(service, issuer, Clock.systemUTC(), Ticker.systemTicker());
+    }
+
+    @Override
+    public EuropeanHcpIdwsToken xcaSoapHeaderToIdwsToken(
+            String soapHeader,
+            String audience
+    ) throws AuthenticationException {
+        // Can't use simple get because of checked exception and lambda.
+        var cacheKey = IdwsCacheKey.fromSoapHeader(soapHeader, audience, issuer);
+        var cached = idwsCache.getIfPresent(cacheKey);
+        if (cached == null) {
+            cached = service.xcaSoapHeaderToIdwsToken(soapHeader, audience);
+            idwsCache.put(cacheKey, cached);
+        }
+        return cached;
+    }
+
+    @Override
+    public DgwsAssertion nspDgwsIdentityToAssertion(NspDgwsIdentity identity) throws AuthenticationException {
+        // Can't use simple get because of checked exception and lambda.
+        var cacheKey = identity;
+        var cached = dgwsCache.getIfPresent(cacheKey);
+        if (cached == null) {
+            cached = service.nspDgwsIdentityToAssertion(identity);
+            dgwsCache.put(cacheKey, cached);
+        }
+        return cached;
+    }
+
+    private record IdwsCacheKey(
+            String audience,
+            String nameIdFormat,
+            String nameId,
+            Set<CacheAttribute> attributes
+    ) {
+        static IdwsCacheKey fromSoapHeader(
+                String soapHeader,
+                String audience,
+                String issuer
+        ) throws AuthenticationException {
+            var ncpAssertions = OpenNcpAssertions.fromSoapHeader(soapHeader);
+            var bstParams = BootstrapTokenParams.fromOpenNcpAssertions(ncpAssertions, audience, issuer);
+            return new IdwsCacheKey(
+                    audience,
+                    bstParams.nameIdFormat(),
+                    bstParams.nameId(),
+                    bstParams.attributes().stream().map(CacheAttribute::new).collect(Collectors.toSet()));
+        }
+    }
+
+    private record CacheAttribute(BootstrapTokenParams.SamlAttribute att) {
+        @Override
+        public boolean equals(Object o) {
+            if (o == null || getClass() != o.getClass()) return false;
+            var that = (CacheAttribute) o;
+            return switch (att) {
+                case BootstrapTokenParams.SamlAttribute.New n1 ->
+                        that.att instanceof BootstrapTokenParams.SamlAttribute.New n2
+                                && Objects.equals(n1.name(), n2.name()) && n1.values().equals(n2.values());
+                case BootstrapTokenParams.SamlAttribute.Raw r1 ->
+                        that.att instanceof BootstrapTokenParams.SamlAttribute.Raw r2
+                                && normalized(r1.node()).isEqualNode(normalized(r2.node()));
```

Comment:
I don't think you hit this code in the unit tests.  Could that be done by fiddling a bit with the soap header?